### PR TITLE
Feature/carousel component

### PR DESCRIPTION
**Pull request by constanzamorillomeli**
## Description
Se agrego el componente carousel.

## Frontify component link
[Frontify - Carousel] (https://company-161429.frontify.com/d/kxHCRixezmfK/n-a#/componentes/carousel)

### APK
https://drive.google.com/file/d/1PuuLbILsJgf1TyJAVdC2C34jJC4DRPGK/view?usp=sharing

## Checks
Are you sure you followed all those steps? In such case, let's say with me "I solemnly declare that ...":
- [x] I have coded an example inside the Demo App,
- [x] I have added proper tests,
- [ ] I have modified the [Changelog](https://github.com/mercadolibre/fury_andesui-android/blob/develop/CHANGELOG.md) file,
- [ ] I have updated GitHub wiki,
- [x] I have the explicit approval of one or more members of the UX Team,
- [ ] I have updated the version in gradle.properties file.

## Change type
This is easy, let us know what this code is about:
- [x] This code adds a new component
- [ ] This code includes improvements to an existent component
- [ ] This code improves or modifies ONLY the Demo App.

## Check common errors
Whether you are an Android or iOS developer and if you're still there then we'll give you a present:
https://proandroiddev.com/how-to-maximize-androids-ui-reusability-5-common-mistakes-cb2571216a9f

### CODE REVIEW

**Comment by furyGorgoryAdmin** [comment]
### Informative Menu <h2>  ### 

### Use any of these tags in a pull request description or leave a comment ### 
| Tag | Description |
|---|---|
| **[canary]**| Creates a build that can only be deployed with canary strategy. | 
| **[build]** |  Creates version or build from a branch that has been tagged as "no releasable". | 
| **[name:version-name]** |Specify the name of the build. For example: [build] [name:1.2.3-name] or [canary] [name:1.2.3-name].|
| **[major]** | To automatically increase the first position of the version. For example: [build] [major] or [canary] [major].|
| **[minor]** | To automatically increase the second position of the version. For example: [build] [minor] or [canary] [minor]. |
| **[patch]** | To automatically increase the third position of the version. For example: [build] [patch] or [canary] [patch].|  

[For more information about ci-cd](http://docs.furycloud.io/?lang=esp&section=businessassurance)
[For more information about semver](https://semver.org/#summary)

**Comment by barriosnahuel** [comment]
Hi @constanzamorillomeli! Welcome to the Andes project 🎉!
Thank you for your first contribution, and thank you for being a new member of the Andes Community 😃

File: components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+     */
+    private fun setupViewId() {
+        if (id == NO_ID) { // If this view has no id
+            id = View.generateViewId()
+        }
+    }
+
+    private fun createConfig() = AndesCarouselConfigurationFactory.create(context, andesCarouselAttrs)
+
+    private fun getPaddingRecyclerView() = (context.resources.displayMetrics.widthPixels * 0.10).toInt()
```
Comment (jperezgonzal) [inline components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt:158]: Esto no lo podemos poner directamente en dimens?

File: components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+import com.mercadolibre.android.andesui.carousel.utils.PaddingItemDecoration
+import com.mercadolibre.android.andesui.carousel.utils.ViewHolderListener
+
+class AndesCarousel : ConstraintLayout {
+
+    private lateinit var andesCarouselAttrs: AndesCarouselAttrs
+    private lateinit var recyclerViewComponent: RecyclerView
+    private lateinit var adapterView: CustomAdapter
+    private lateinit var viewManager: LinearLayoutManager
+    private var dataSet: List<Any> = emptyList()
```
Comment (jperezgonzal) [inline components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt:27]: Esto lo podemos llevar para andesCarouselAttrs?

File: components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+import com.mercadolibre.android.andesui.carousel.padding.AndesCarouselPadding
+import com.mercadolibre.android.andesui.carousel.utils.CustomAdapter
+import com.mercadolibre.android.andesui.carousel.utils.PaddingItemDecoration
+import com.mercadolibre.android.andesui.carousel.utils.ViewHolderListener
+
+class AndesCarousel : ConstraintLayout {
+
+    private lateinit var andesCarouselAttrs: AndesCarouselAttrs
+    private lateinit var recyclerViewComponent: RecyclerView
+    private lateinit var adapterView: CustomAdapter
```
Comment (jperezgonzal) [inline components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt:25]: Le cambiaría el nombre a esto 

`private lateinit var carouselAdapter: CarouselAdapter`

File: demoapp/src/main/res/drawable/rounded_corner.xml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android">
+
+    <stroke android:width="7dp" android:color="@color/andes_blue_ml_300" />
+    <corners android:radius="40dp" />
+
+</shape>
```
Comment (jperezgonzal) [inline demoapp/src/main/res/drawable/rounded_corner.xml:7]: Podemos eliminar esto y hacerlo por código?

File: demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/feature/ProgressShowcaseActivity.kt [hunk 5]
```diff
@@ -34,7 +34,7 @@ class ProgressShowcaseActivity : AppCompatActivity() {
         indicator.attach(viewPager)
 
         val adapter = viewPager.adapter as AndesShowcasePagerAdapter
-        //addProgress(adapter.views[0])
+        // addProgress(adapter.views[0])
```
Comment (jperezgonzal) [inline demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/feature/ProgressShowcaseActivity.kt:37]: Se que esto es del Progress y seguramente salto con el lint, pero capaz que ya lo podemos dejar 😄 
Podemos eliminar la variable "adapter" porque no se esta usando ¿no?

File: demoapp/src/main/res/drawable/rounded_corner.xml (hunk 4 above)
Comment (constanzamorillomeli) [reply demoapp/src/main/res/drawable/rounded_corner.xml:7]: Si obvio tincho, esto es para lo que seria la testapp. :)

File: components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt (hunk 3 above)
Comment (constanzamorillomeli) [reply components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt:25]: Dale me parece correcto.

File: demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/feature/ProgressShowcaseActivity.kt (hunk 5 above)
Comment (constanzamorillomeli) [reply demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/feature/ProgressShowcaseActivity.kt:37]: ah mira! Después aprovecho y lo saco

File: components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt (hunk 1 above)
Comment (constanzamorillomeli) [reply components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt:158]: @jperezgonzal no se me ocurre como ponerlo en dimens tincho, pense mas en un calculo dinámico dependiendo las dimensiones del dispositivo por eso lo hice ahi. Me comentas como la pensaste vos? :)

File: components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+class AndesCarousel : ConstraintLayout {
+
+    private lateinit var andesCarouselAttrs: AndesCarouselAttrs
+    private lateinit var recyclerViewComponent: RecyclerView
+    private lateinit var viewManager: LinearLayoutManager
+    private lateinit var _delegate: AndesCarouselDelegate
+
+    /**
+     * Getter and setter for [type].
```
Comment (jperezgonzal) [inline components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt:29]: Actualizar la doc

File: components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    var padding: AndesCarouselPadding
+        get() = andesCarouselAttrs.andesCarouselPadding
+        set(value) {
+            andesCarouselAttrs = andesCarouselAttrs.copy(andesCarouselPadding = value)
+            setupPaddingRecyclerView(createConfig())
+        }
+
+    var delegate: AndesCarouselDelegate
```
Comment (jperezgonzal) [inline components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt:38]: Capaz que podemos agregar la doc aca?
Y capaz que en lugar de **_delegate** podemos usar **andesDelegate** ¿que opinas?

File: components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            id = View.generateViewId()
+        }
+    }
+
+    private fun createConfig() = AndesCarouselConfigurationFactory.create(context, andesCarouselAttrs)
+
+    private fun getPaddingRecyclerView() = (context.resources.displayMetrics.widthPixels * PERCENTAGE).toInt()
+
+    companion object {
+        const val PERCENTAGE = 0.10
```
Comment (jperezgonzal) [inline components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt:134]: Esto es el porcentaje que se ven de la cards consiguientes? Vamos a permitir que sea custom?

File: components/src/main/res/layout/andes_layout_carousel.xml [hunk 9]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="utf-8"?>
+<android.support.constraint.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (jperezgonzal) [inline components/src/main/res/layout/andes_layout_carousel.xml:2]: Es necesario tener el layout? 
AndesCarousel no puede extender directamente de un RecyclerView?

File: components/src/test/java/com/mercadolibre/android/andesui/carousel/AndesCarouselTest.kt [hunk 10]
```diff
@@ -0,0 +1,14 @@
+package com.mercadolibre.android.andesui.carousel
+
+import com.mercadolibre.android.andesui.carousel.factory.AndesCarouselAttrs
+import com.mercadolibre.android.andesui.carousel.factory.AndesCarouselConfigurationFactory
+import org.mockito.Mockito
+import org.robolectric.RuntimeEnvironment
+
+class AndesCarouselTest {
```
Comment (jperezgonzal) [inline components/src/test/java/com/mercadolibre/android/andesui/carousel/AndesCarouselTest.kt:8]: Esto esta pendiente no?

File: components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt (hunk 8 above)
Comment (constanzamorillomeli) [reply components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt:134]: No dijeron nada de que sea custom, dijeron que mostremos un porcentaje de item siguiente. Podemos preguntar si le damos la posibilidad al usuario o no.

File: components/src/test/java/com/mercadolibre/android/andesui/carousel/AndesCarouselTest.kt (hunk 10 above)
Comment (constanzamorillomeli) [reply components/src/test/java/com/mercadolibre/android/andesui/carousel/AndesCarouselTest.kt:8]: Sisi asi es :D

File: components/src/main/res/layout/andes_layout_carousel.xml (hunk 9 above)
Comment (constanzamorillomeli) [reply components/src/main/res/layout/andes_layout_carousel.xml:2]: No me lo permite Tincho 👎  dice que por tema de superclass no esta permitido. **"cannot access recyclerview which is a supertype of . . ."** @jperezgonzal

File: components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt (hunk 6 above)
Comment (constanzamorillomeli) [reply components/src/main/java/com/mercadolibre/android/andesui/carousel/AndesCarousel.kt:29]: Dale

File: components/src/test/java/com/mercadolibre/android/andesui/carousel/AndesCarouselTest.kt (hunk 10 above)
Comment (constanzamorillomeli) [reply components/src/test/java/com/mercadolibre/android/andesui/carousel/AndesCarouselTest.kt:8]: Me voy a poner

**Review by jperezgonzal — COMMENTED** [review]
@constanzamorillomeli te animas a subir un gif de la demoapp? 🙏

**Comment by constanzamorillo1** [comment]
> ### Thanks for starting a pull request on Andes UI!
> ## Description
> Se agrego el componente carousel.
> 
> ## Frontify component link
> [Frontify - Carousel] (https://company-161429.frontify.com/d/kxHCRixezmfK/n-a#/componentes/carousel)
> 
> ## Screenshots
> 
> 
> 
> 
> ### APK
> https://drive.google.com/file/d/1HNfr4JGjj5RXtuchUnngoVMRVwoMI7Qj/view?usp=sharing
> 
> ## Checks
> Are you sure you followed all those steps? In such case, let's say with me "I solemnly declare that ...":
> 
> * [x]  I have coded an example inside the Demo App,
> * [ ]  I have added proper tests,
> * [ ]  I have modified the [Changelog](https://github.com/mercadolibre/fury_andesui-android/blob/develop/CHANGELOG.md) file,
> * [ ]  I have updated GitHub wiki,
> * [x]  I have the explicit approval of one or more members of the UX Team,
> * [ ]  I have updated the version in gradle.properties file.
> 
> ## Change type
> This is easy, let us know what this code is about:
> 
> * [x]  This code adds a new component
> * [ ]  This code includes improvements to an existent component
> * [ ]  This code improves or modifies ONLY the Demo App.
> 
> ## Check common errors
> Whether you are an Android or iOS developer and if you're still there then we'll give you a present:
> https://proandroiddev.com/how-to-maximize-androids-ui-reusability-5-common-mistakes-cb2571216a9f

**Review by jperezgonzal — APPROVED** [review]

File: demoapp/src/main/res/values/strings.xml [hunk 11]
```diff
@@ -64,13 +64,15 @@
     <string name="tag_description">Tags are compact components that represent complex information in a summarized way. Allowing users to make selections, enter information, or filter content.</string>
     <string name="snackbar_description">They communicate quickly and concisely the status of an event. For more complex or extensive communications, consider the use of tooltip or message.</string>
     <string name="checkbox_description">They allow multiple selections from a list of options, unlike radio buttons that only allow you to choose a single option.</string>
+    <string name="carousel_description">Groups and allows you to manipulate a horizontal sequence of content.</string>
```
Comment (snti) [inline demoapp/src/main/res/values/strings.xml:67]: El texto parece estar incompleto

Comment (constanzamorillomeli) [reply demoapp/src/main/res/values/strings.xml:67]: que texto santi? No entendi

**Review by snti — APPROVED** [review]

**Review by snti — DISMISSED** [review]

**Review by RodriP — APPROVED** [review]

**Review by jperezgonzal — APPROVED** [review]

**Review by snti — APPROVED** [review]

**Pull request merged**